Refactor ActiveSupport::XMLMini test suite

maclover7 opened the pull request:
### Summary

As I am working on getting Active Support's test suite running on JRuby, I noticed many inconsistencies with the test suite for `ActiveSupport::XMLMini`, including many tests missing from certain adapters. This PR aims to standardize the tests run for each of the different `XMLMini` backends. Also, I noticed that tests for the libxml engine were not running locally due to a missing gem. I added the gem to the Gemfile so all future contributors will be able to run the full suite without needing to install another gem.

maclover7 committed:
Do not include `ActiveSupport` into test classes

Will help get rid of errors like the following:

```
  1) Error:
JDOMEngineTest#test_order=:
ArgumentError: wrong number of arguments (0 for 1)
    /Users/jon/code/rails/rails/activesupport/lib/active_support/core_ext/module/attribute_accessors.rb:106:in `test_order='
```

maclover7 committed:
Install libxml-ruby gem so full XMLMini test suite can be run

maclover7 committed:
Create `CommonXMLMiniAdapterTest`; common tests module

Created a new module (a la Action Cable subscription adapter's test
suite) to be included in all sub class test to ensure compatability and
reduce duplicated code.

kaspth commented on activesupport/test/xml_mini/libxml_engine_test.rb:
```diff
@@ -6,198 +6,24 @@
   require "abstract_unit"
   require "active_support/xml_mini"
   require "active_support/core_ext/hash/conversions"
+  require_relative "./common"
```
Why do we need `./` when we require relatively to the current dir?

maclover7 committed:
Code styling cleanup

kaspth commented on activesupport/test/xml_mini/common.rb:
```diff
@@ -0,0 +1,229 @@
+module CommonXMLMiniAdapterTest
```
Since all but JDOM engine tests are of the form:

```ruby
begin
  require "libxml"
rescue LoadError
  # Skip libxml tests
else
  require "abstract_unit"
  require "active_support/xml_mini"
  require "active_support/core_ext/hash/conversions"
  require_relative "./common"

  class LibxmlEngineTest < ActiveSupport::TestCase
  end
end
```

I think I'd abstract this as an custom `TestCase` instead:

```ruby
# xml_mini/engine_test.rb
require "abstract_unit"
require "active_support/xml_mini"
require "active_support/core_ext/hash/conversions"

class XMLMiniEngineTest < ActiveSupport::TestCase
  def self.run_with_gem(gem_name)
    require gem_name
    yield
  rescue LoadError # Skip tests
  end

  setup do
    @old_engine, XMLMini.backend = XMLMini.backend, engine
  end
  
  teardown do
    XMLMini.backend = @old_engine
  end
  
  # All the standard tests...
end

# xml_mini/libxml_engine_test.rb
require_relative 'xml_mini_engine_test'

XMLMiniEngineTest.skip_unless_gem("libxml") do
  class LibxmlEngineTest < XMLMiniEngineTest
    private
      # I dislike the `adapter_name` name since the nomenclature is `engine` :)
      def engine
        "LibXML"
      end
      
      def expansion_error_attack
        # ...
      end
  end
end
```

kaspth commented on activesupport/test/xml_mini/common.rb:
```diff
@@ -0,0 +1,229 @@
+
+    def jdom_skip
+      skip "JDOM does not support this test" if adapter_name == "JDOM"
+    end
```
I'd think of the methods skipped on `JDOM` as extended engine tests — because for the common module or class to know about JDOM seems off — and would abstract like this:

```ruby
def test_skipped
  require_extended_engine

  assert_equal_rexml # yada yada...
end

private
  def require_extended_engine
    skip "#{engine} is just a basic engine" unless extended_engine?
  end

  def extended_engine? # Override in the JDOM tests and return false.
    true
  end
end
```

The `extended` term could stand to improve, same with the `skip` message.

kaspth reviewed (approved):
Overall this looks good to me and improves these tests a ton. Most of my comments are just about how to take these tests even further, in my opinion. 🖖

maclover7 committed:
Create XMLMiniEngineTest; base class for all tests

maclover7 committed:
Move tests to a module, include in subclasses, style linting

This fixes an error where the test runner would try and run
XMLMiniEngineTest like a normal test class, except it's abstract. Now,
to circumvent this, we don't include any of the actual tests in
XMLMiniEngineTest; they are wrapped in a module that is included in
subclass when they inherit from XMLMiniEngineTest. Pretty neat, huh?

kaspth commented on activesupport/test/xml_mini/xml_mini_engine_test.rb:
```diff
@@ -0,0 +1,260 @@
+  end
+
+  def self.run_with_platform(platform_name)
+    yielf if RUBY_PLATFORM.include?(platform_name)
```
`yield` is what you want. Though I'd just keep the code in the file :)

kaspth commented on activesupport/test/xml_mini/xml_mini_engine_test.rb:
```diff
@@ -0,0 +1,260 @@
+  def self.inherited(base)
+    base.class_eval do
+      include EngineTests
+    end
```
I think we can just call `include` directly on `base`.

kaspth commented on activesupport/test/xml_mini/xml_mini_engine_test.rb:
```diff
@@ -0,0 +1,260 @@
+    def teardown
+      ActiveSupport::XmlMini.backend = @default_backend
+      super
+    end
```
`setup` and `teardown` can live in the base class just fine. Then I think we can also strip the `ActiveSupport::` namespace like they were in the original test files.

maclover7 replied:
👍 on moving `setup` and `teardown`, but we've got to keep the `ActiveSupport::` namespace or else we'll run into this:

```
 98) Error:
REXMLEngineTest#test_parse_from_frozen_string:
NameError: uninitialized constant XMLMiniEngineTest::XmlMini
    /Users/jon/code/rails/rails/activesupport/test/xml_mini/xml_mini_engine_test.rb:26:in `setup'
```

kaspth commented on activesupport/test/xml_mini/xml_mini_engine_test.rb:
```diff
@@ -0,0 +1,260 @@
+          <member>
+          &a;
+          </member>
+      eoxml
```
Indent `eoxml` 2 spaces :)

kaspth commented:
All right, feel free to squash the commits. Then I'll merge if the build passes 👏

maclover7 commented:
👍 cut number of commits in half, and build is passing (not sure why Code Climate isn't updating on the PR, but [their build](https://codeclimate.com/github/rails/rails/pull/27111#builds) is passing)
